### PR TITLE
sync_diff_inspector: add check report && context && README

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
1. add check report, will be print at the end of the log.
2. add context control
3. add readme
4. some minor update

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@GregoryIan  @july2993 PTAL

File: sync_diff_inspector/README.md [hunk 1]
```diff
@@ -1 +1,30 @@
 # sync-diff-inspector
+
+sync-diff-inspector is a tool for compare two database's data.
```
Comment (IANTHEREAL) [inline sync_diff_inspector/README.md]: comparing?

File: sync_diff_inspector/chunk.go [hunk 2]
```diff
@@ -272,21 +266,15 @@ func GenerateCheckJob(db *sql.DB, Schema string, table *model.TableInfo, splitFi
...
 		}
-		if reflect.TypeOf(chunk.begin).Kind() == reflect.Int64 {
-			where = fmt.Sprintf("(`%s` %s %d AND `%s` %s %d)", column.Name, gt, chunk.begin, column.Name, lt, chunk.end)
-		} else if reflect.TypeOf(chunk.begin).Kind() == reflect.Float64 {
-			where = fmt.Sprintf("(`%s` %s %f AND `%s` %s %f)", column.Name, gt, chunk.begin, column.Name, lt, chunk.end)
-		} else {
-			where = fmt.Sprintf("(`%s` %s \"%v\" AND `%s` %s \"%v\")", column.Name, gt, chunk.begin, column.Name, lt, chunk.end)
-		}
-		where = fmt.Sprintf("%s AND %s", where, limitRange)
+		where := fmt.Sprintf("(`%s` %s ? AND `%s` %s ? AND %s)", column.Name, gt, column.Name, lt, limitRange)
```
Comment (GregoryIan) [inline sync_diff_inspector/chunk.go]: does GetRandomValues always return string?

File: pkg/dbutil/common.go [hunk 3]
```diff
@@ -133,18 +133,17 @@ func GetRowCount(ctx context.Context, db *sql.DB, schemaName string, tableName s
 	return cnt.Int64, nil
 }
 
-// GetRandomValues returns some random value of a column, not used for number type column.
-// TODO: why not used for number type column? fix it
-func GetRandomValues(ctx context.Context, db *sql.DB, schemaName string, table string, column string, num int64, min, max interface{}, limitRange string) ([]string, error) {
+// GetRandomValues returns some random value of a column.
+func GetRandomValues(ctx context.Context, db *sql.DB, schemaName, table, column string, num int64, min, max interface{}, limitRange string) ([]string, error) {
```
Comment (IANTHEREAL) [inline pkg/dbutil/common.go]: does GetRandomValues always return string?

File: sync_diff_inspector/report.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	DataEqual   bool
+}
+
+// Report saves the check results.
+type Report struct {
+	sync.RWMutex
+
+	Schema         string
+	TableNameEqual bool
+	SourceTables   []string
```
Comment (IANTHEREAL) [inline sync_diff_inspector/report.go]: need them?

File: sync_diff_inspector/report.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	return &Report{
+		Schema:         schema,
+		TableResults:   make(map[string]*TableResult),
+		TableNameEqual: true,
+		Pass:           true,
+	}
+}
+
+// String returns a string of this Report.
+func (r *Report) String() (report string) {
```
Comment (IANTHEREAL) [inline sync_diff_inspector/report.go:57]: let fail item display firstly

File: sync_diff_inspector/report.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	sync.RWMutex
+
+	Schema         string
+	TableNameEqual bool
+	SourceTables   []string
+	TargetTables   []string
+	PassNum        int32
+	FailedNum      int32
+	TableResults   map[string]*TableResult
+	Pass           bool
```
Comment (IANTHEREAL) [inline sync_diff_inspector/report.go]: just use `Pass` and `Fail` to represent it

File: sync_diff_inspector/report.go (hunk 5 above)
Comment (IANTHEREAL) [reply sync_diff_inspector/report.go:57]: give a display example

File: pkg/dbutil/common.go (hunk 3 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/common.go]: return interface{}

File: sync_diff_inspector/report.go (hunk 5 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/report.go:57]: done

Comment (IANTHEREAL) [reply sync_diff_inspector/report.go:57]: If u have time, u can use a template to do it that is easier to modify and maintain

Comment (IANTHEREAL) [reply sync_diff_inspector/report.go:57]: do u forget read lock in this function?

File: pkg/dbutil/common.go [hunk 7]
```diff
@@ -133,25 +133,24 @@ func GetRowCount(ctx context.Context, db *sql.DB, schemaName string, tableName s
 	return cnt.Int64, nil
 }
 
-// GetRandomValues returns some random value of a column, not used for number type column.
-// TODO: why not used for number type column? fix it
-func GetRandomValues(ctx context.Context, db *sql.DB, schemaName string, table string, column string, num int64, min, max interface{}, limitRange string) ([]string, error) {
+// GetRandomValues returns some random value of a column.
+func GetRandomValues(ctx context.Context, db *sql.DB, schemaName, table, column string, num int64, min, max interface{}, limitRange string) ([]interface{}, error) {
```
Comment (IANTHEREAL) [inline pkg/dbutil/common.go:137]: any tests after u modify `getRandomValues`?

File: sync_diff_inspector/diff.go [hunk 8]
```diff
@@ -122,26 +126,35 @@ func (df *Diff) Equal() (equal bool, err error) {
...
+		eq2, err := df.EqualTableData(table)
 		if err != nil {
 			log.Errorf("equal table error %v", err)
+			return errors.Trace(err)
 		}
-		if !eq {
-			equal = false
+		if !eq2 {
 			log.Errorf("table %s's data is not equal", table.Name)
+			df.report.Result = Fail
```
Comment (IANTHEREAL) [inline sync_diff_inspector/diff.go]: do it in `SetTableDataCheckResult` or `SetTableStructCheckResult`?

**Comment by IANTHEREAL** [comment]
Rest LGTM

File: sync_diff_inspector/report.go (hunk 5 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/report.go:57]: add read lock done. and I will use template next pr.

File: pkg/dbutil/common.go (hunk 7 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/common.go:137]: I have already test.

**Comment by WangXiangUSTC** [comment]
@july2993  PTAL

File: sync_diff_inspector/report.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+		table: test2
+		table's struct equal
+		table's data not equal
+
+		table: test3
+		table's struct equal
+		table's data equal
+	*/
+	report = fmt.Sprintf("\ncheck result of schema %s: %s!\n", r.Schema, r.Result)
+	report = fmt.Sprintf("%s%d tables' check passed, %d tables' check failed.\n", report, r.PassNum, r.FailedNum)
```
Comment (july2993) [inline sync_diff_inspector/report.go]: use bytes.Buffer or just `report += ***` may be more clear

File: sync_diff_inspector/main.go [hunk 10]
```diff
@@ -52,47 +52,55 @@ func main() {
...
 	}
 
-	ok, err := d.Equal()
+	err = d.Equal()
 	if err != nil {
 		log.Fatalf("check data difference error %v", errors.ErrorStack(err))
 	}
 
-	return ok
+	log.Info(d.report.String())
```
Comment (july2993) [inline sync_diff_inspector/main.go]: no need String()

**Review by july2993 — APPROVED** [review]
LGTM

**Pull request merged**